tests/v2/validation/airgap/k3s_custom_cluster_test.go

```diff
@@ -0,0 +1,146 @@
+)
+
+type AirGapK3SCustomClusterTestSuite struct {
+	suite.Suite
+	client             *rancher.Client
+	session            *session.Session
+	kubernetesVersions []string
+	cnis               []string
+	corralImage        string
+	corralAutoCleanup  bool
+	registryFQDN       string
+	advancedOptions    provisioning.AdvancedOptions
+}
+
+func (a *AirGapK3SCustomClusterTestSuite) TearDownSuite() {
+	a.session.Cleanup()
+}
+
+func (a *AirGapK3SCustomClusterTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	a.session = testSession
+
+	clustersConfig := new(provisioning.Config)
+	config.LoadConfig(provisioning.ConfigurationFileKey, clustersConfig)
+
+	corralRancherHA := new(corralha.CorralRancherHA)
+	config.LoadConfig(corralha.CorralRancherHAConfigConfigurationFileKey, corralRancherHA)
+
+	registriesConfig := new(registries.Registries)
+	config.LoadConfig(registries.RegistriesConfigKey, registriesConfig)
+
+	a.kubernetesVersions = clustersConfig.K3SKubernetesVersions
+	a.cnis = clustersConfig.CNIs
+	a.advancedOptions = clustersConfig.AdvancedOptions
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(a.T(), err)
+
+	a.client = client
+	listOfCorrals, err := corral.ListCorral()
+	require.NoError(a.T(), err)
+
+	corralConfig := corral.CorralConfigurations()
+
+	err = corral.SetupCorralConfig(corralConfig.CorralConfigVars, corralConfig.CorralConfigUser, corralConfig.CorralSSHPath)
+	require.NoError(a.T(), err)
+
+	corralPackage := corral.CorralPackagesConfig()
+	a.corralImage = corralPackage.CorralPackageImages[corralPackageAirgapCustomClusterName]
+	a.corralAutoCleanup = corralPackage.Cleanup
+
+	_, corralExist := listOfCorrals[corralRancherHA.Name]
+	if corralExist {
+		bastionIP, err := corral.GetCorralEnvVar(corralRancherHA.Name, corralRegistryIP)
+		require.NoError(a.T(), err)
+
+		err = corral.UpdateCorralConfig(corralBastionIP, bastionIP)
+		require.NoError(a.T(), err)
+
+		registryFQDN, err := corral.GetCorralEnvVar(corralRancherHA.Name, corralRegistryFQDN)
+		require.NoError(a.T(), err)
+		logrus.Infof("registry fqdn is %s", registryFQDN)
+		a.registryFQDN = registryFQDN
+	} else {
+		a.registryFQDN = registriesConfig.ExistingNoAuthRegistryURL
+	}
+
+}
+
+func (a *AirGapK3SCustomClusterTestSuite) TestProvisioningK3SCustomCluster() {
+	nodeRoles := map[int]string{
+		1: "--etcd --controlplane --worker",
+	}
+
+	var name string
+	for _, kubeVersion := range a.kubernetesVersions {
+		name = "Airgap Provision K3S Custom Cluser Kubernetes version: " + kubeVersion
+		for _, cni := range a.cnis {
+			name += " cni: " + cni
+			a.Run(name, func() {
+				clusterName := testProvisionAirgapRKE2K3SCustomCluster(a.T(), a.client, nodeRoles, a.corralImage, cni, kubeVersion, a.corralAutoCleanup, a.advancedOptions)
+				checkPodStatusImageSource(a.T(), a.client, clusterName, a.registryFQDN)
+			})
+		}
+	}
+}
+
+func (a *AirGapK3SCustomClusterTestSuite) TestProvisioningUpgradeK3SCustomCluster() {
```

Comment:
will there be a `dynamic` func similar to what we have in provisioning/k3s?

Reply:
I wasn't planning on it, since in the release testing we really don't define those. If you think it would be useful sure. Just one note, because of how the corral package works the nodeRoles is slightly different that what is in provisioning.

Reply:
okay lets leave it as-is for now. if we need it later, there can be separate work done instead.